feat: breed look for mate

brainsmarts opened the pull request:
context: added an action that allows a creature to look for a suitable mate, what determines it for now is simply if they have enough energy, but traits can adjust their strategy

update:
- small rename of Time Manager variable for less confusion
- changed TimeManager from start to awake so that creatures can use the TimeManager when first created
- Added a look for mate action, it only looks for a suitable mate does not breed with the found mate

testing:
tested in normal sample scene, with various creatures to see if any mates are found, and if the mates were found in the correct condition

The target may need to be passed on in the future, which may be as simple as having a target in the creature's data

brainsmarts committed:
feat: breed look for mate

context: added an action that allows a creature to look for a suitable mate, what determines it for now is simply if they have enough energy, but traits can adjust their strategy

update:
- small rename of Time Manager variable for less confusion
- changed TimeManager from start to awake so that creatures can use the TimeManager when first created
- Added a look for mate action, it only looks for a suitable mate does not breed with the found mate

testing:
tested in normal sample scene, with various creatures to see if any mates are found, and if the mates were found in the correct condition

The target may need to be passed on in the future, which may be as simple as having a target in the creature's data

brainsmarts committed:
feat: breeding

context: after finishing LookForMate, create a breeding action
creature looks for some valid creature to breed, either currently looking to breed or looking for mate, then roomba moves towards each other, then both creatures has a 50% each to create a creature

update:
- updated createcreature to correct speed
- added breeding action
- added a ToString to LookForMate
- added Breeding to the base creature actions

testing:
simple testing in sample scene with debug.log, over time with 5x speed creatures kept multiplying without any lag or crashes

Kornfuzion commented on My project/Assets/Scripts/CreatureLogic/Actions/Breeding/Breed.cs:
```diff
@@ -0,0 +1,122 @@
+    //decrease energy
+    public void OnExit()
+    {
+        if(Random.Range(0f,1f) > .5)
```
what is this? can we make it clearer what this condition is - or potentially pull it out into a new function?

brainsmarts replied:
changed .5 to the variable child chance, its the chance of spawning a child

Kornfuzion commented on My project/Assets/Scripts/CreatureLogic/Actions/Breeding/Breed.cs:
```diff
@@ -0,0 +1,122 @@
+    RangeScanner scanner;
+    Rigidbody2D rb;
+    float breedDuration;
+    float breedTimeEnd;
+    float nextBreedTime;
```
can we simplify this by removing nextBreedTime and having it be derived from breedTimeEnd + breedCooldown?

Kornfuzion commented on My project/Assets/Scripts/CreatureLogic/Actions/Breeding/Breed.cs:
```diff
@@ -0,0 +1,122 @@
+    //else do nothing
+    public void Run()
+    {
+        if (target.currentActionNode.action.ToString().Equals("Breed"))
+        {
+            //Debug.Log("Can Breed");
+            canBreed = true;
+        }
```
my thought is that this should be closer to the evaluation on line 85. otherwise we might check that canBreed=true which falls out of sync with the actual target's breed state when we're checking proximity.

If it makes sense, i'd be fine with leaving Run() empty and putting all of this logic into EndCondition

Kornfuzion commented on My project/Assets/Scripts/CreatureLogic/Actions/Breeding/LookForMate.cs:
```diff
@@ -0,0 +1,108 @@
+    int requiredEnergy;
+    //the time when creature can look for mate again
+    //ie: time is 1:20 then creature can only look for mate after 1:20
+    float nextLookForMate = 0;
```
same comment as above: remove this variable and derive it from searchTimeEnd + cooldown

Kornfuzion commented on My project/Assets/Scripts/CreatureLogic/Actions/Breeding/LookForMate.cs:
```diff
@@ -0,0 +1,108 @@
+using System.Collections.Generic;
+using UnityEngine;
+
+public class LookForMate : IAction
```
this whole class seems kind of redundant, as above in Breed, we check if a target is either in Breed already or LookForMate. can we consolidate them into one action - specifically, just use Breed for both of these usecases?

brainsmarts replied:
I agree, this problem also happens with find food and eat food, so to fix this I believe its better to finally revamp the creature data so things like findfood and eatfood share the same food variable and findmate and breed have the same target as was discussed earlier, I'll change this in the next update

Kornfuzion reviewed (approved):
approving to unblock you. my gut feeling is we can get away with a single breed action instead of two. i'll let you think about that and see if we can simplify things

brainsmarts committed:
update: added the changes from the comments

update:
- removed nextXTime from both breed and findmate
- used a variable for child chance
- moved currentActionNode == breed to the end check instead of the run check
